### PR TITLE
new recipe: splunk-opentelemetry-cpp

### PR DESCRIPTION
**Pull request by seemk**
**splunk-opentelemetry-cpp/0.4.0**

Hi, this is a first to-be-published version of [Splunk's distribution of OpenTelemetry for C++](https://github.com/signalfx/splunk-otel-cpp). This won't be a dependency of other libraries we intend to package. I'm one of the devs for the distribution and the goal is to ease the usage process.

Any suggestions welcome, thanks :slightly_smiling_face: 

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by seemk** [comment]
It looks the includes are borked, fixing it

**Comment by seemk** [comment]
@madebr @uilianries Updated/rewrote the recipe to use `opentelemetry/1.0.1` as a dependency. Waiting on https://github.com/conan-io/conan-center-index/pull/7799 now

**Comment by seemk** [comment]
The Windows job seems to get stuck every time :thinking:

**Comment by uilianries** [comment]
Let's restart.

**Comment by seemk** [comment]
@uilianries Perhaps I should recreate this PR?

**Comment by uilianries** [comment]
@seemk we have had an internal problem, open a new PR won't solve this situation. Conan team is working to fix it asap.

**Comment by seemk** [comment]
> @seemk we have had an internal problem, open a new PR won't solve this situation. Conan team is working to fix it asap.

Thanks, appreciated!

**Comment by seemk** [comment]
@uilianries Is there anything that I can do to help here?

**Comment by madebr** [comment]
It looks like Windows is hanging.
Perhaps the test is failing there?
The Linux logs show:
```
----Running------
> bin/test_package
-----------------
[Error] File: /home/conan/w/BuildSingleReference/.conan/data/opentelemetry-cpp/1.0.1/_/_/build/68e1659b012a357550b3f136fa4a0741e68f62d1/source_subfolder/exporters/otlp/src/otlp_grpc_exporter.cc:149[OTLP Exporter] Export() failed: failed to connect to all addresses
```

**Comment by seemk** [comment]
> It looks like Windows is hanging. Perhaps the test is failing there? The Linux logs show:
> 
> ```
> ----Running------
> > bin/test_package
> -----------------
> [Error] File: /home/conan/w/BuildSingleReference/.conan/data/opentelemetry-cpp/1.0.1/_/_/build/68e1659b012a357550b3f136fa4a0741e68f62d1/source_subfolder/exporters/otlp/src/otlp_grpc_exporter.cc:149[OTLP Exporter] Export() failed: failed to connect to all addresses
> ```

Thanks, this message is expected though (default conf tries to export via OTLP) and the program exits normally. I can't see the Windows logs, so no idea what's going on there :slightly_frowning_face:

**Comment by madebr** [comment]
Looks like it does not support dll's.
You can fix it in 2 ways: a dirty way or the proper way.
- Proper way: patch the header to add `__declspec(dllexport)`/`__declspec(dllimport)` to the correct spaces
It looks like this is needed in 2 places: [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L26) and [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L45-L46).
[The cmake `GenerateExportHeader`](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) might be useful for this.
- Dirty way: add `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)` to the cmake wrapper.

**Comment by prince-chrismc** [comment]
My vote is the dirty way :broom: :bucket:

**Comment by seemk** [comment]
> Looks like it does not support dll's. You can fix it in 2 ways: a dirty way or the proper way.
> 
> * Proper way: patch the header to add `__declspec(dllexport)`/`__declspec(dllimport)` to the correct spaces
>   It looks like this is needed in 2 places: [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L26) and [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L45-L46).
>   [The cmake `GenerateExportHeader`](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) might be useful for this.
> * Dirty way: add `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)` to the cmake wrapper.

Thanks! I changed the library to use `GenerateExportHeader`.

> It looks like Windows is hanging. Perhaps the test is failing there? The Linux logs show:
> 
> ```
> ----Running------
> > bin/test_package
> -----------------
> [Error] File: /home/conan/w/BuildSingleReference/.conan/data/opentelemetry-cpp/1.0.1/_/_/build/68e1659b012a357550b3f136fa4a0741e68f62d1/source_subfolder/exporters/otlp/src/otlp_grpc_exporter.cc:149[OTLP Exporter] Export() failed: failed to connect to all addresses
> ```

And you were right about this one, I think it's an issue with OpenTelemetry C++ on Windows where it's waiting indefinitely to flush the spans :thinking: and the export timeouts for OTLP exporters are not working (even though they are configurable). I've changed it to Jaeger exporter which does have proper timeouts.


Should be all good now, thanks for the help everyone. @madebr @prince-chrismc

**Review by madebr — APPROVED** [review]

**Comment by madebr** [comment]
> > Looks like it does not support dll's. You can fix it in 2 ways: a dirty way or the proper way.
> > 
> > * Proper way: patch the header to add `__declspec(dllexport)`/`__declspec(dllimport)` to the correct spaces
> >   It looks like this is needed in 2 places: [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L26) and [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L45-L46).
> >   [The cmake `GenerateExportHeader`](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) might be useful for this.
> > * Dirty way: add `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)` to the cmake wrapper.
> 
> Thanks! I changed the library to use `GenerateExportHeader`.

Great! Using this cmake module and not hand rolling this also makes it a more smooth experience for your users.
The downside is that usages of the same headers and switching between static/shared splunk-opentelemetry-cpp libraries is impossible now (as is the case on linux systems): the headers fix the library type.

> > It looks like Windows is hanging. Perhaps the test is failing there? The Linux logs show:
> > ```
> > ----Running------
> > > bin/test_package
> > -----------------
> > [Error] File: /home/conan/w/BuildSingleReference/.conan/data/opentelemetry-cpp/1.0.1/_/_/build/68e1659b012a357550b3f136fa4a0741e68f62d1/source_subfolder/exporters/otlp/src/otlp_grpc_exporter.cc:149[OTLP Exporter] Export() failed: failed to connect to all addresses
> > ```
> 
> And you were right about this one, I think it's an issue with OpenTelemetry C++ on Windows where it's waiting indefinitely to flush the spans thinking and the export timeouts for OTLP exporters are not working (even though they are configurable). I've changed it to Jaeger exporter which does have proper timeouts.
> 
> Should be all good now, thanks for the help everyone. @madebr @prince-chrismc

Running your library on a new platform often gives surprises and headaches.
I'm happy you were able to fix it and and up with a better library.

File: recipes/splunk-opentelemetry-cpp/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+from conans import ConanFile, CMake, tools
+from conans.errors import ConanInvalidConfiguration
+import os
+import glob
+
+
+class SplunkOpentelemetryConan(ConanFile):
+    name = "splunk-opentelemetry-cpp"
+    license = "Apache 2.0"
```
Comment (prince-chrismc) [inline recipes/splunk-opentelemetry-cpp/all/conanfile.py:9]: https://spdx.org/licenses/Apache-2.0.html

```suggestion
    license = "Apache-2.0"
```

File: recipes/splunk-opentelemetry-cpp/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    def _remove_unnecessary_package_files(self):
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
```
Comment (prince-chrismc) [inline recipes/splunk-opentelemetry-cpp/all/conanfile.py:51]: There's no `pkg_config` target declared in `package_info`

**Review by prince-chrismc — COMMENTED** [review]
Super tiny comments. It looks really good!!!

File: recipes/splunk-opentelemetry-cpp/all/conanfile.py (hunk 2 above)
Comment (seemk) [reply recipes/splunk-opentelemetry-cpp/all/conanfile.py]: Ah thanks, the `pkgconfig` deletion was a remnant from an earlier recipe version. Removed it now

**Review by madebr — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Comment by seemk** [comment]
@uilianries Think this needs your powers here

**Review by danimtb — APPROVED** [review]

**Comment by prince-chrismc** [comment]
Ping since the PR is blocked

#7565 | seemk | Oct 6 | splunk-opentelemetry-cpp | 25 | Nov 24 | @uilianries | madebr, prince-chrismc, danimtb
-- | -- | -- | -- | -- | -- | -- | --

**Comment by seemk** [comment]
@uilianries :pray:

**Comment by prince-chrismc** [comment]
Today might be his first day back from vacation based on his status. Shouldn't be too long

**Review by SSE4 — APPROVED** [review]

**Comment by prince-chrismc** [comment]
Ping @uilianries since the PR is blocked

**Review by uilianries — APPROVED** [review]
LGTM sorry my delay

**Pull request merged**